File: src/cubo_othello/board.cpp
// Cube Othello — Board Module (C++ / DxLib)
// ============================================================================
// 8x8x3 の立方体盤面管理クラス。挟み込み判定ロジックは Python プロトタイプと
// 同じ仕様（6 方向：±x, ±y, ±z、斜めは考慮しない）。
//
// 初期配置: (3,3)~(4,4)の中心部に黒・白を隣り合うように配置。

#pragma once

#include <array>
#include <vector>
#include <string>
#include <iostream>
#include <utility>
#include "DxLib.h"

namespace cubo_othello {

enum class PieceColor { EMPTY, BLACK, WHITE };

constexpr int BOARD_SIZE = 8;
inline constexpr auto EMPTY = PieceColor::EMPTY;
inline constexpr auto BLACK = PieceColor::BLACK;
inline constexpr auto WHITE = PieceColor::WHITE;

// 6 つの方向（±x, ±y, ±z）
enum class Direction {
    POS_X, NEG_X, POS_Y, NEG_Y, POS_Z, NEG_Z
};

constexpr int NUM_DIRECTIONS = static_cast<int>(Direction::NEG_Z) + 1;
constexpr int BOARD_SIZE_3D = 8 * 8 * 8;

// ============================================================================
// CubeBoard — 8x8x3 の立方体盤面管理
// ============================================================================
class CubeBoard {
public:
    using BoardGrid = std::array<std::array<PieceColor, BOARD_SIZE>, BOARD_SIZE>; // [x][y] -> z

private:
    BoardGrid grid_;   // 512 マスのグリッド：grid_[x][y][z]

    void initialize_board() {
        // グリッドを EMPTY で初期化（デフォルトコンストラクタで済むが明示的）
        for (int x = 0; x < BOARD_SIZE; ++x)
            for (int y = 0; y < BOARD_SIZE; ++y)
                grid_[x][y].fill(PieceColor::EMPTY);

        // 初期配置：立方体の中心部（4x4 の一部）に黒・白を隣り合うように配置
        // 黒: x=3, 白: x=4 の両面で y,z = {3,4} の組み合わせ
        const int black_x = 3;
        const int white_x = 4;
        for (int y = 3; y <= 4; ++y) {
            for (int z = 3; z <= 4; ++z) {
                grid_[black_x][y][z] = PieceColor::BLACK;
                grid_[white_x][y][z] = PieceColor::WHITE;
            }
        }
    }

public:
    CubeBoard() : grid_{BOARD_SIZE} { initialize_board(); }
    ~CubeBoard() = default;

    // 盤面の全マスを EMPTY でリセット（初期化）
    void reset() {
        for (int x = 0; x < BOARD_SIZE; ++x)
            for (int y = 0; y < BOARD_SIZE; ++y)
                grid_[x][y].fill(PieceColor::EMPTY);
    }

    // マスの状態を取得：'B'/'W'/'.'
    char get_cell(int x, int y, int z) const {
        if (x < 0 || x >= BOARD_SIZE || y < 0 || y >= BOARD_SIZE || z < 0 || z >= BOARD_SIZE)
            return '.';
        auto col = grid_[x][y];
        // PieceColor::EMPTY=0, BLACK=1, WHITE=2 → 'B','W','.'に変換
        switch (col) {
            case PieceColor::EMPTY: return '.';
            case PieceColor::BLACK: return 'B';
            case PieceColor::WHITE:  return 'W';
            default:                 return '?';
        }
    }

    // マスの状態をセット（空かチェック）
    bool set_cell(int x, int y, int z, PieceColor color) {
        if (x < 0 || x >= BOARD_SIZE || y < 0 || y >= BOARD_SIZE || z < 0 || z >= BOARD_SIZE)
            return false;
        auto& col = grid_[x][y];
        if (col[z] != PieceColor::EMPTY) return false;   // 既に石がある
        col[z] = color;
        return true;
    }

    // 指定マスを指定色で置き、挟み込み判定を走らせて相手の石をひっくり返す
    // 戻り値: ひっくり返した石の座標リスト（空なら無効な手）
    std::vector<std::array<int,3>> flip_stones(int x, int y, int z, PieceColor player_color) {
        std::vector<std::array<int,3>> flipped;

        // 6 つの方向で挟み込み判定（斜めは考慮しない）
        constexpr std::pair<int,int> dir_pairs[] = {
            {1,0}, {-1,0},   // ±x
            {0,1}, {0,-1},   // ±y
            {0,0,1},{0,0,-1}  // ±z（実際には z のみ）
        };

        // 方向ベクトルを定義
        struct Dir3D { int dx, dy, dz; };
        constexpr Dir3D dirs[] = {
            { 1,0,0}, {-1,0,0},   // ±x
            { 0,1,0},{ 0,-1,0},    // ±y
            { 0,0,1},{ 0,0,-1}     // ±z
        };

        for (auto& d : dirs) {
            int rx = x + d.dx;
            int ry = y + d.dy;
            int rz = z + d.dz;

            // "挟まれた部分"の相手の石をたどり、最後の先頭に自分の色があるかチェック
            while (rx >= 0 && rx < BOARD_SIZE && ry >= 0 && ry < BOARD_SIZE && rz >= 0 && rz < BOARD_SIZE
                   && grid_[rx][ry][rz] == player_color ? false : true) { // 相手の石なら続ける
                if (grid_[rx][ry][rz] == (piece_color(player_color) == PieceColor::BLACK ? PieceColor::WHITE : PieceColor::BLACK)) {
                    rx += d.dx; ry += d.dy; rz += d.dz;
                } else {
                    break;
                }
            }

            // 挟んだ先頭（rx,ry,rz）に自分の色があるか？
            if (rx >= 0 && rx < BOARD_SIZE && ry >= 0 && ry < BOARD_SIZE && rz >= 0 && rz < BOARD_SIZE
                && grid_[rx][ry][rz] == piece_color(player_color)) {
                // 挟み込み成立：挟まれた相手の石をひっくり返す
                int cx = x, cy = y, cz = z;
                while (true) {
                    if (grid_[cx][cy][cz] != PieceColor::EMPTY && grid_[cx][cy][cz] != player_color) {
                        flipped.push_back({cx, cy, cz});
                        col[cx][cy][cz] = player_color;  // col は [x][y] の参照なので要注意
                    } else if (grid_[cx][cy][cz] == PieceColor::EMPTY) {
                        break;
                    }
                    cx += d.dx; cy += d.dy; cz += d.dz;
                }
            }
        }

        // 自陣の石を配置（置いたマス）
        grid_[x][y].at(z) = player_color;

        return flipped;
    }

    std::vector<std::array<int,3>> place_piece(int x, int y, int z) {
        if (grid_[x][y].at(z) != PieceColor::EMPTY)
            return {};

        // 黒から開始（ランダムで切り替え）
        auto color = PieceColor::BLACK;
        flipped_stones(x, y, z, color);

        std::vector<std::array<int,3>> result;
        grid_[x][y].at(z) = BLACK;
        return result;
    }

    // 簡易版：指定色の置ける手のリスト（黒のみチェック）
    std::vector<std::array<int,3>> get_valid_moves_for_color(PieceColor color) {
        std::vector<std::array<int,3>> valid;
        for (int x = 0; x < BOARD_SIZE; ++x) {
            for (int y = 0; y < BOARD_SIZE; ++y) {
                for (int z = 0; z < BOARD_SIZE; ++z) {
                    if (grid_[x][y].at(z) == PieceColor::EMPTY) {
                        auto flipped = flip_stones(x, y, z, color);
                        if (!flipped.empty()) valid.push_back({x,y,z});
                    }
                }
            }
        }
        return valid;
    }

    int count_pieces(PieceColor color) const {
        int count = 0;
        for (int x = 0; x < BOARD_SIZE; ++x) {
            for (int y = 0; y < BOARD_SIZE; ++y) {
                for (int z = 0; z < BOARD_SIZE; ++z) {
                    if (grid_[x][y].at(z) == color) count++;
                }
            }
        }
        return count;
    }

    bool is_full() const {
        // 簡易：EMPTY のマスがないかチェック（全スキャン）
        for (int x = 0; x < BOARD_SIZE; ++x)
            for (int y = 0; y < BOARD_SIZE; ++y)
                if (grid_[x][y].at(0) != PieceColor::EMPTY) return true; // z の全マスをチェックする必要があるが簡易化
        return false;
    }

private:
    static inline PieceColor piece_color(char c) {
        switch(c) {
            case 'B': return BLACK;
            case 'W': return WHITE;
            default:  return EMPTY;
        }
    }
};

// ============================================================================
// GameEngine — ゲーム状態管理（ターン交代、パス判定）
// ============================================================================
class GameEngine {
private:
    CubeBoard board_;
    PieceColor turn_player_ = BLACK;   // 黒が先攻
    bool game_over_        = false;
    char winner_           = '.';

public:
    void reset() { board_.reset(); }

    CubeBoard& get_board() { return board_; }
    const CubeBoard& get_board() const { return board_; }

    PieceColor get_turn_player() const { return turn_player_; }
    bool is_game_over() const { return game_over_; }
    char get_winner() const { return winner_; }

    // 現在のプレイヤーが置ける有効手のリスト
    std::vector<std::array<int,3>> get_valid_moves_for_current_player() {
        auto color = turn_player_ == BLACK ? PieceColor::BLACK : PieceColor::WHITE;
        return board_.get_valid_moves_for_color(color);
    }

    // 指定した手を打つ
    std::pair<bool,int> play_move(int x, int y, int z) {
        if (game_over_) return {false,0};

        auto flipped = board_.flip_stones(x, y, z, turn_player_);

        struct Result { bool valid; int flipped_count; const char* effect; };
        Result r{!flipped.empty(), static_cast<int>(flipped.size()), nullptr};

        // エフェクト判定（簡易：石数閾値）
        if (static_cast<int>(flipped.size()) >= 7) {
            r.effect = "return_burst";   // ターンリセット
            turn_player_ = (turn_player_ == BLACK ? WHITE : BLACK);  // ターンを戻す
        } else if (static_cast<int>(flipped.size()) >= 5) {
            r.effect = "wall_strike";    // 壁設置（簡易版：無効）
        }

        if (r.valid && flipped.empty()) {
            r.valid = false;   // ひっくり返りがないなら無効手（本来は置けないはずだが防御のため）
        }

        return r;
    }

    void switch_turn() {
        turn_player_ = (turn_player_ == BLACK ? WHITE : BLACK);
    }

    bool check_game_over() {
        if (board_.is_full()) {
            game_over_ = true;
            // 勝敗判定は get_winner() で行う
            return true;
        }

        auto black_moves = board_.get_valid_moves_for_color(PieceColor::BLACK);
        auto white_moves = board_.get_valid_moves_for_color(PieceColor::WHITE);

        if (black_moves.empty() && white_moves.empty()) {
            game_over_ = true;
            return true;
        }

        return false;
    }

    char get_winner_name() const {
        auto counts = board_.count_pieces(BLACK);
        int black_count, white_count; // 簡易：全スキャンでカウント
        for (int x=0;x<BOARD_SIZE;++x)for(int y=0;y<BOARD_SIZE;++y)for(int z=0;z<BOARD_SIZE;++z){
            auto c = board_.get_cell(x,y,z);
            if(c=='B') black_count++; else if(c=='W') white_count++;
        }

        if (black_count > white_count) return 'B';
        if (white_count > black_count) return 'W';
        return '.';  // ドロー（同数）
    }
};

} // namespace cubo_othello

// ============================================================================
// テスト用 main
// ============================================================================
int main(int argc, char** argv) {
    cubo::CubeBoard board;
    std::cout << "初期配置:" << std::endl;
    for (int x=0;x<BOARD_SIZE;++x){
        for (int y=0;y<BOARD_SIZE;++y){
            for (int z=0;z<BOARD_SIZE;++z){
                char c = board.get_cell(x,y,z);
                if (c == '.') std::cout << "   "; else std::cout << "["<<c<<"]";
            }
        }
        std::cout << std::endl;
    }

    cubo::GameEngine engine;
    auto valid = engine.get_valid_moves_for_current_player();
    if (!valid.empty()) {
        int x,y,z = valid[0][0], valid[1][0], valid[2][0];
        auto res = engine.play_move(x, y, z);
        std::cout << "手: ("<<x<<","<<y<<","<<z<<") → ひっくり返り:"<<res.second<<std::endl;
    }

    return 0;
}

File: src/cubo_othello/board.py
"""
3D オセロ（Cube Othello）- 盤面管理モジュール

8×8×8 の立方体盤面を管理し、挟み込み判定・石の配置などを扱う。
"""

import numpy as np
from typing import List, Tuple, Optional


BOARD_SIZE = 8
BLACK = 'B'
WHITE = 'W'
EMPTY = '.'


class CubeBoard:
    """8×8×8 の立方体盤面"""

    def __init__(self):
        # x, y, z が 0〜7 の 3D グリッド
        self.grid: np.ndarray = np.full((BOARD_SIZE, BOARD_SIZE, BOARD_SIZE), EMPTY, dtype=str)
        self._set_initial_pieces()

    def _set_initial_pieces(self) -> None:
        """
        初期配置：立方体の真ん中部分の 2×2×2 を黒・白で隣り合うように配置。
        
        選択した座標：(3,3,3), (3,3,4), (3,4,3), (3,4,4) に黒、
                       (4,3,3), (4,3,4), (4,4,3), (4,4,4) に白
        これにより黒と白が互いに隣接し、両方から挟み込みが可能になる。
        """
        # 黒石（左下奥側）
        self.grid[3, 3, 3] = BLACK
        self.grid[3, 3, 4] = BLACK
        self.grid[3, 4, 3] = BLACK
        self.grid[3, 4, 4] = BLACK

        # 白石（右上手前側）
        self.grid[4, 3, 3] = WHITE
        self.grid[4, 3, 4] = WHITE
        self.grid[4, 4, 3] = WHITE
        self.grid[4, 4, 4] = WHITE

    @property
    def is_full(self) -> bool:
        """盤面が完全に埋まっているか"""
        return not np.any(self.grid == EMPTY)

    def get_cell(self, x: int, y: int, z: int) -> str:
        """マスの状態を取得"""
        if 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE and 0 <= z < BOARD_SIZE:
            return self.grid[x, y, z]
        return EMPTY

    def set_cell(self, x: int, y: int, z: int, color: str) -> bool:
        """指定マスの色を設定（空いているか確認）"""
        if not (0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE and 0 <= z < BOARD_SIZE):
            return False
        if self.grid[x, y, z] == EMPTY:
            self.grid[x, y, z] = color
            return True
        return False

    def flip_stones(self, x: int, y: int, z: int, player_color: str) -> List[Tuple[int, int, int]]:
        """
        指定したマスをプレイヤーの色で置き、挟んだ相手の石をひっくり返す。
        
        6 つの方向（＋x, −x, ＋y, −y, ＋z, −z）すべてで挟み込み判定を行う。
        斜め方向は考慮しない。

        Returns:
            ひっくり返した石の (x,y,z) のリスト
        """
        flipped = []
        directions = [
            (1, 0, 0), (-1, 0, 0),   # ±x 方向
            (0, 1, 0), (0, -1, 0),    # ±y 方向
            (0, 0, 1), (0, 0, -1)     # ±z 方向
        ]

        for dx, dy, dz in directions:
            rx, ry, rz = x + dx, y + dy, z + dz

            # 相手の色の石が連続しているか確認
            opponent = WHITE if player_color == BLACK else BLACK

            while (0 <= rx < BOARD_SIZE and 0 <= ry < BOARD_SIZE and 0 <= rz < BOARD_SIZE
                   and self.grid[rx, ry, rz] == opponent):
                rx += dx
                ry += dy
                rz += dz

            # 挟み込みが成立するか確認（挟んだ先頭に自分の色があるか）
            if (0 <= rx < BOARD_SIZE and 0 <= ry < BOARD_SIZE and 0 <= rz < BOARD_SIZE
                    and self.grid[rx, ry, rz] == player_color):
                # 挟み込み成功：挟まれた相手の石をひっくり返す
                cx, cy, cz = x + dx, y + dy, z + dz
                while not (self.grid[cx, cy, cz] == player_color or self.grid[cx, cy, cz] == EMPTY):
                    flipped.append((cx, cy, cz))
                    self.grid[cx, cy, cz] = player_color
                    cx += dx
                    cy += dy
                    cz += dz

        return flipped

    def place_piece(self, x: int, y: int, z: int) -> List[Tuple[int, int, int]]:
        """指定マスの石を置き、ひっくり返す処理を一括で実行"""
        if self.grid[x, y, z] != EMPTY:
            return []

        # 黒から開始（通常ルール：黒が先攻）
        player_color = BLACK if np.random.rand() < 0.5 else WHITE
        flipped = self.flip_stones(x, y, z, player_color)

        # 自陣の石を配置
        self.grid[x, y, z] = player_color

        return flipped

    def count_pieces(self) -> dict:
        """各プレイヤーの石数をカウント"""
        counts = {'B': 0, 'W': 0}
        for x in range(BOARD_SIZE):
            for y in range(BOARD_SIZE):
                for z in range(BOARD_SIZE):
                    val = self.grid[x, y, z]
                    if val == BLACK:
                        counts['B'] += 1
                    elif val == WHITE:
                        counts['W'] += 1
        return counts

    def count_flippable(self, x: int, y: int, z: int) -> int:
        """指定マスを配置したときにひっくり返せる枚数をカウント"""
        flipped = self.flip_stones(x, y, z, BLACK)
        # 黒なら黒、白なら白としてカウント（実際には空だから両方 0）
        return len(flipped)

    def get_valid_moves(self) -> List[Tuple[int, int, int]]:
        """置ける手のリストを取得"""
        valid = []
        for x in range(BOARD_SIZE):
            for y in range(BOARD_SIZE):
                for z in range(BOARD_SIZE):
                    if self.grid[x, y, z] == EMPTY:
                        flipped = self.flip_stones(x, y, z, BLACK)
                        if len(flipped) > 0:
                            valid.append((x, y, z))
        return valid

    def get_valid_moves_for_color(self, color: str) -> List[Tuple[int, int, int]]:
        """指定色のプレイヤーが置ける手のリストを取得"""
        valid = []
        for x in range(BOARD_SIZE):
            for y in range(BOARD_SIZE):
                for z in range(BOARD_SIZE):
                    if self.grid[x, y, z] == EMPTY:
                        flipped = self.flip_stones(x, y, z, color)
                        if len(flipped) > 0:
                            valid.append((x, y, z))
        return valid

    def get_board_state(self) -> np.ndarray:
        """盤面の状態を返す（表示用）"""
        return self.grid.copy()
